crates/red_knot_module_resolver/src/path.rs

```diff
@@ -0,0 +1,1132 @@
+    pub(crate) fn with_pyi_extension(&self) -> Self {
+        ModuleResolutionPathRef::from(self).with_pyi_extension()
+    }
+
+    #[must_use]
+    pub(crate) fn with_py_extension(&self) -> Option<Self> {
+        ModuleResolutionPathRef::from(self).with_py_extension()
+    }
+
+    #[must_use]
+    pub(crate) fn relativize_path<'a>(
+        &'a self,
+        absolute_path: &'a (impl AsRef<FileSystemPath> + ?Sized),
+    ) -> Option<ModuleResolutionPathRef<'a>> {
+        ModuleResolutionPathRef::from(self).relativize_path(absolute_path.as_ref())
+    }
+
+    /// Returns `None` if the path doesn't exist, isn't accessible, or if the path points to a directory.
+    pub(crate) fn to_vfs_file(
+        &self,
+        db: &dyn Db,
+        search_path: &Self,
+        typeshed_versions: &LazyTypeshedVersions,
+        target_version: TargetVersion,
+    ) -> Option<VfsFile> {
+        ModuleResolutionPathRef::from(self).to_vfs_file(
+            db,
+            search_path,
+            typeshed_versions,
+            target_version,
+        )
+    }
+}
+
+impl fmt::Debug for ModuleResolutionPathBuf {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let (name, path) = match &self.0 {
+            ModuleResolutionPathBufInner::Extra(path) => ("Extra", path),
+            ModuleResolutionPathBufInner::FirstParty(path) => ("FirstParty", path),
+            ModuleResolutionPathBufInner::SitePackages(path) => ("SitePackages", path),
+            ModuleResolutionPathBufInner::StandardLibrary(path) => ("StandardLibrary", path),
+        };
+        f.debug_tuple(&format!("ModuleResolutionPathBuf::{name}"))
+            .field(path)
+            .finish()
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Hash, Clone, Copy)]
+enum ModuleResolutionPathRefInner<'a> {
+    Extra(&'a FileSystemPath),
+    FirstParty(&'a FileSystemPath),
+    StandardLibrary(&'a FileSystemPath),
+    SitePackages(&'a FileSystemPath),
+}
+
+impl<'a> ModuleResolutionPathRefInner<'a> {
+    #[must_use]
+    fn query_stdlib_version(
+        module_path: &'a FileSystemPath,
+        typeshed_versions: &LazyTypeshedVersions,
+        stdlib_search_path: Self,
+        stdlib_root: &FileSystemPath,
+        db: &dyn Db,
+        target_version: TargetVersion,
+    ) -> TypeshedVersionsQueryResult {
+        let Some(module_name) = stdlib_search_path
+            .relativize_path(module_path)
+            .and_then(Self::to_module_name)
+        else {
+            return TypeshedVersionsQueryResult::DoesNotExist;
+        };
+        typeshed_versions.query_module(&module_name, db, stdlib_root, target_version)
+    }
+
+    #[must_use]
+    fn is_directory(
+        &self,
+        db: &dyn Db,
+        search_path: Self,
+        typeshed_versions: &LazyTypeshedVersions,
+        target_version: TargetVersion,
+    ) -> bool {
+        match (self, search_path) {
+            (Self::Extra(path), Self::Extra(_)) => db.file_system().is_directory(path),
```

Comment:
I really don't like that we ignore the passed-in search path entirely for non-stdlib variants in so many of these methods. It seems very error-prone. At the very least I'd want to see some debug assertions of consistency, though I'd rather have an API that didn't have redundant information in the first place (e.g. if module paths were enforced as always relative and required to be combined with a search path to be turned into an absolute path.)